tools/cell_census_builder/validate.py

```diff
@@ -276,59 +276,86 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
-                    )
-            if actual_x_coo is not None:
-                expected_ad_x_coo = expected_ad_x.tocoo()
-                assert expected_ad_x_coo.nnz == actual_x_coo.nnz, error.format(
-                    "nnz", actual_x_coo.nnz, expected_ad_x.nnz
-                )
-                assert expected_ad_x_coo.shape == actual_x_coo.shape, error.format(
-                    "shape", actual_x_coo.shape, expected_ad_x_coo.shape
+                expected_ad_x, expected_ad_var = ad.raw.X, ad.raw.var
+
+            # get the joinids for the obs axis
+            obs_joinids = (
+                exp.obs.read(
+                    column_names=["soma_joinid", "dataset_id"], value_filter=f"dataset_id == '{dataset.dataset_id}'"
                 )
-                assert (
-                    expected_ad_x_coo != actual_x_coo
-                ).nnz == 0, f"{eb.name}:{dataset.dataset_id} the X matrix elements are not equal."
-            expected_gene_count = np.count_nonzero(expected_ad_x.sum(axis=0))
-            assert expected_gene_count == actual_census_gene_count, error.format(
-                "gene_count", actual_census_gene_count, expected_gene_count
+                .concat()
+                .to_pandas()
+            )
+
+            assert ad.n_obs == len(obs_joinids)
+            if len(obs_joinids) == 0:
+                continue
+
+            X_raw = (
+                exp.ms[MEASUREMENT_RNA_NAME]
+                .X["raw"]
+                .read((obs_joinids.soma_joinid.to_numpy(), slice(None)))
+                .tables()
+                .concat()
+            )
+            data = X_raw["soma_data"].to_numpy()
+            rows = X_raw["soma_dim_0"].to_numpy()
+            cols = X_raw["soma_dim_1"].to_numpy()
+
+            # positionally (re)index obs/rows. We _know_ that the Census assigns
+            # obs soma_joinids in the obs position order of the original AnnData, so
+            # leverage that for simplicity and speed.
+            rows_by_position = obs_joinids.set_index("soma_joinid").index.get_indexer(rows)
+
+            # get the joinids for the var axis
+            all_var_ids = (
+                exp.ms[MEASUREMENT_RNA_NAME].var.read(column_names=["soma_joinid", "feature_id"]).concat().to_pandas()
             )
+            # mask defines which feature_ids are in the AnnData
+            var_joinid_in_adata = all_var_ids.feature_id.isin(expected_ad_var.index)
+            assert ad.n_vars == var_joinid_in_adata.sum()
+
+            # positionally re-index the vars/cols using the feature_id as the join key.
+            cols_by_position = (
+                expected_ad_var.join(all_var_ids.set_index("feature_id"))
+                .set_index("soma_joinid")
+                .index.get_indexer(cols)
+            )
+
+            # Assertion 1 - the contents of the X matrix are EQUAL for all var values
+            # present in the AnnData
+            assert (
```

Comment:
I [read](https://stackoverflow.com/questions/30685024/check-if-two-scipy-sparse-csr-matrix-are-equal) that both shape and nnz should be equal before doing this comparison. Shapes makes sense. However I don't see how asserting nnz help.

Reply:
You may be misunderstanding the code -- this an equality test implemented with `__ne__` and an nnz check.  This pattern prevents the equality check from having to materialize an full array of bools (remember, the equality check is per-element).

See for example, this [stack overflow answer](https://stackoverflow.com/questions/30685024/check-if-two-scipy-sparse-csr-matrix-are-equal).  

Highly suggest you study the sparse array implementation and documentation to get a better understanding of how it works.